main.go

```diff
@@ -0,0 +1,382 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/pingcap/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type config []struct {
+	Targets []string `yaml:"targets"`
+}
+
+func main() {
+	err := validateEnvironmentVariables()
+	if err != nil {
+		log.WithError(err).Error("Environment variable validation failed")
+		err = sendMattermostErrorNotification(err, "Environment variable validation failed")
+		if err != nil {
+			log.WithError(err).Error("Failed to send Mattermost error notification")
+		}
+		return
+	}
+
+	err = thanosStoreDiscovery()
+	if err != nil {
+		log.WithError(err).Error("Failed to run Thanos store discovery")
+		err = sendMattermostErrorNotification(err, "The Thanos store discovery failed")
+		if err != nil {
+			log.WithError(err).Error("Failed to send Mattermost error notification")
+		}
+		return
+	}
+}
+
+// validateEnvironmentVariables is used to validate the environment variables needed by Thanos store discovery.
+func validateEnvironmentVariables() error {
+	privateHostedZoneID := os.Getenv("PRIVATE_HOSTED_ZONE_ID")
+	if len(privateHostedZoneID) == 0 {
+		return errors.Errorf("PRIVATE_HOSTED_ZONE_ID environment variable is not set")
+	}
+
+	thanosNamespace := os.Getenv("THANOS_NAMESPACE")
+	if len(thanosNamespace) == 0 {
+		return errors.Errorf("THANOS_NAMESPACE environment variable is not set")
+	}
+
+	thanosDeploymentName := os.Getenv("THANOS_DEPLOYMENT_NAME")
+	if len(thanosDeploymentName) == 0 {
+		return errors.Errorf("THANOS_DEPLOYMENT_NAME environment variable is not set.")
+	}
+
+	thanosConfigMapName := os.Getenv("THANOS_CONFIGMAP_NAME")
+	if len(thanosConfigMapName) == 0 {
+		return errors.Errorf("THANOS_CONFIGMAP_NAME environment variable is not set.")
+	}
+
+	mattermostAlertsHook := os.Getenv("MATTERMOST_ALERTS_HOOK")
+	if len(mattermostAlertsHook) == 0 {
+		return errors.Errorf("MATTERMOST_ALERTS_HOOK environment variable is not set.")
+	}
+	return nil
+}
+
+// thanosStoreDiscovery is used to keep Thanos up to date with deployed Thanos Query targets.
+func thanosStoreDiscovery() error {
+	privateHostedZoneID := os.Getenv("PRIVATE_HOSTED_ZONE_ID")
+	thanosNamespace := os.Getenv("THANOS_NAMESPACE")
+	thanosDeploymentName := os.Getenv("THANOS_DEPLOYMENT_NAME")
+	thanosConfigMapName := os.Getenv("THANOS_CONFIGMAP_NAME")
```

Comment:
Instead of re-reading these env vars here it would be a bit cleaner to read them in a single place, possibly changing `validateEnvironmentVariables` to `validateAndGetEnvVars` that returns a struct with all the env vars in it, and then pass them in as an argument to `thanosStoreDiscovery`.

If left as is there is a couple issues that could arise:

1. It's easy to change the env var name in one place and not another by accident, introducing a bug
2. There's actually a tiny, though unlikely, chance for a race here. If somehow the env vars on the system change/gets unset between the running `validateEnvironmentVariables` and `thanosStoreDiscovery` then you could get some weird or bad behaviour

Reply:
All validation is moved in a new function that validates and returns a structure that can be used by the thanosStoreDiscovery.